File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/config/builder/EngineConfig.kt
/*
 * Copyright 2017-2020 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.docs.config.builder

// tag::imports[]
import io.micronaut.context.annotation.ConfigurationBuilder
import io.micronaut.context.annotation.ConfigurationProperties
// end::imports[]

// tag::class[]
@ConfigurationProperties("my.engine") // <1>
internal class EngineConfig {
    @ConfigurationBuilder(prefixes = ["with"])  // <2>
    val builder = EngineImpl.builder()

    @ConfigurationBuilder(prefixes = ["with"], configurationPrefix = "crank-shaft") // <3>
    val crankShaft = CrankShaft.builder()

    @set:ConfigurationBuilder(prefixes = ["with"], configurationPrefix = "spark-plug") // <4>
    var sparkPlug = SparkPlug.builder()
}
// end::class[]

File: test-suite-python/src/test/python/micronaut/docs/config/builder/EngineConfig.py
from typing import Annotated

from .CrankShaft import CrankShaft
from .CrankShaft import CrankShaftBuilder
from .EngineImpl import EngineImpl
from .EngineImpl import EngineImplBuilder
from .SparkPlug import SparkPlug
from .SparkPlug import SparkPlugBuilder

# tag::imports[]
from micronaut.context.annotation import ConfigurationBuilder
from micronaut.context.annotation import ConfigurationProperties
# end::imports[]


# tag::class[]
@ConfigurationProperties("my.engine")  # <1>
class EngineConfig:
    builder: Annotated[EngineImplBuilder, ConfigurationBuilder(prefixes="with")] = EngineImpl.builder()  # <2>

    crank_shaft: Annotated[
        CrankShaftBuilder,
        ConfigurationBuilder(prefixes="with", configurationPrefix="crank-shaft"),
    ] = CrankShaft.builder()  # <3>

    _spark_plug: SparkPlugBuilder = SparkPlug.builder()

    @property
    def spark_plug(self) -> SparkPlugBuilder:
        return self._spark_plug

    @spark_plug.setter
    @ConfigurationBuilder(prefixes="with", configurationPrefix="spark-plug")  # <4>
    def spark_plug(self, spark_plug: SparkPlugBuilder) -> None:
        self._spark_plug = spark_plug
# end::class[]
